# Проект по SQL

Компания решила купила крупный сервис для чтения книг по подписке. Наша первая задача  — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Описание данных
<br>

Таблица books
<br>
Содержит данные о книгах:
<br>
book_id — идентификатор книги;
<br>
author_id — идентификатор автора;
<br>
title — название книги;
<br>
num_pages — количество страниц;
<br>
publication_date — дата публикации книги;
<br>
publisher_id — идентификатор издателя.
<br>

Таблица authors
<br>
Содержит данные об авторах:
<br>
author_id — идентификатор автора;
<br>
author — имя автора.
<br>

Таблица publishers
<br>
Содержит данные об издательствах:
<br>
publisher_id — идентификатор издательства;
<br>
publisher — название издательства;
<br>

Таблица ratings
<br>
Содержит данные о пользовательских оценках книг:
<br>
rating_id — идентификатор оценки;
<br>
book_id — идентификатор книги;
<br>
username — имя пользователя, оставившего оценку;
<br>
rating — оценка книги.
<br>

Таблица reviews
<br>
Содержит данные о пользовательских обзорах:
<br>
review_id — идентификатор обзора;
<br>
book_id — идентификатор книги;
<br>
username — имя автора обзора;
<br>
text — текст обзора.

Задания:
<br>
- Посчитаем, сколько книг вышло после 1 января 2000 года;
<br>
- Для каждой книги посчитаем количество обзоров и среднюю оценку;
<br>
- Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
<br>
- Определим автора с самой высокой средней оценкой книг — учитем только книги с 50 и более оценками;
<br>
- Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [1]:
! pip install psycopg2-binary 

In [2]:
import pandas as pd
import sqlalchemy as sa
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)


In [3]:
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)


In [4]:
# формируем запрос и выводим данные
query = '''SELECT * FROM books LIMIT 5'''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query = '''SELECT COUNT(*) FROM books'''
get_sql_data(query)

count
0   1000

In [6]:
query = ''' SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_NAME='books' '''
get_sql_data(query)

count
0      6

В таблице books 6 столбцов и 1000 строк

In [7]:
query = '''
SELECT * FROM ratings LIMIT 5
'''
get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
query = '''SELECT COUNT(*) FROM ratings'''
get_sql_data(query)

count
0   6456

In [9]:
query = ''' SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_NAME='ratings' '''
get_sql_data(query)

count
0      4

В таблице ratings 4 столбца и 6456 строк

In [10]:
query = '''
SELECT * FROM authors LIMIT 5
'''
get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [11]:
query = '''SELECT COUNT(*) FROM authors'''
get_sql_data(query)

count
0    636

In [12]:
query = ''' SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_NAME='authors' '''
get_sql_data(query)

count
0      2

В таблице authors 2 столбца и 636 строк

In [13]:
query = '''
SELECT * FROM reviews LIMIT 5
'''
get_sql_data(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [14]:
query = '''SELECT COUNT(*) FROM reviews'''
get_sql_data(query)

count
0   2793

In [15]:
query = ''' SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_NAME='reviews' '''
get_sql_data(query)

count
0      4

В таблице reviews 4 столбца и 2793 строки

In [16]:
query = '''
SELECT * FROM publishers LIMIT 5
'''
get_sql_data(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [17]:
query = '''SELECT COUNT(*) FROM publishers'''
get_sql_data(query)

count
0    340

In [18]:
query = ''' SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_NAME='publishers' '''
get_sql_data(query)

count
0      2

В таблице publishers 2 столбца и 340 строк

##### Задания

Посчитаем, сколько книг вышло после 1 января 2000 года;

In [19]:
query = '''
SELECT count(distinct(book_id)) 
FROM books 
WHERE DATE_TRUNC('day', publication_date)::date > '2000-01-01'
'''
get_sql_data(query)

count
0    819

Всего книг вышедших после 01.01.2000 года 819.

Для каждой книги посчитаем количество обзоров и среднюю оценку;

In [20]:
query = '''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT rv.review_id) AS reviews, 
       ROUND(AVG(r.rating),3) AS avg_rating
FROM books AS b
LEFT JOIN reviews AS rv
ON b.book_id = rv.book_id
LEFT JOIN  ratings AS r
ON b.book_id = r.book_id
GROUP BY b.book_id
ORDER BY reviews DESC
'''
get_sql_data(query)

book_id                                              title  reviews  \
0        948                            Twilight (Twilight  #1)        7   
1        963                                Water for Elephants        6   
2        734                                   The Glass Castle        6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...        6   
4        695  The Curious Incident of the Dog in the Night-Time        6   
..       ...                                                ...      ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel        0   
996      808                            The Natural Way to Draw        0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites        0   
998      221                          Essential Tales and Poems        0   
999      191  Disney's Beauty and the Beast (A Little Golden...        0   

     avg_rating  
0         3.663  
1         3.977  
2         4.207  
3         4.415  
4         4.081  
..          ...  
995       3.667  
996       3.000  
997       5.000  
998       4.000  
999       4.000  

[1000 rows x 4 columns]

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;


In [21]:
query = '''
SELECT  p.publisher AS publisher, 
COUNT(b.book_id) AS books
FROM publishers as p
INNER JOIN books as b ON b.publisher_id = p.publisher_id 
WHERE 
num_pages > 50
GROUP BY 
p.publisher 
ORDER BY 
books DESC 

LIMIT 1;
'''
get_sql_data(query)

publisher  books
0  Penguin Books     42

Издательство, которое выпустило наибольшее число книг : Penguin Books

Определим автора с самой высокой средней оценкой книг — учтем только книги с 50 и более оценками;


In [22]:
query = '''
Select a.author_id as id, a.author as author, avg(r.rating) as avg_rating, count(r.rating) as count_rating
From authors as a
join books as b on b.author_id = a.author_id
join ratings as r on r.book_id = b.book_id
group by id, author
having count(r.rating) >= 50
order by count_rating desc
limit 1
'''
get_sql_data(query)

id                      author  avg_rating  count_rating
0  236  J.K. Rowling/Mary GrandPré    4.288462           312

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré

Средний рейтинг по всем книгам автора, у которых 50 и более обзоров - 4.288462

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [23]:
query = '''
SELECT
AVG(qw.reviews_amount) AS avg_reviews

FROM (SELECT wq.username AS username,
COUNT(r.review_id) AS reviews_amount 
      
FROM (SELECT
rt.username AS username
FROM ratings as rt
GROUP BY rt.username
HAVING COUNT(rt.rating_id) > 48) AS wq 
INNER JOIN reviews as r ON r.username = wq.username

GROUP BY
wq.username
) AS qw;

'''

get_sql_data(query)

avg_reviews
0         24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24

# Вывод: 
В данном проекте были изучены данные сервиса для формирования концепции нового продукта. 
<br>
В ходе исследования мы выяснили:
<br>
1) После 1 января 2000 года вышло 819 книг
<br>
2) Для каждой книги были посчитаны средний рейтинг и число обзоров
<br>
3) Издательство, которое выпустило наибольшее число книг : Penguin Books - 42 книги
<br>
4) Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré. Средний рейтинг по всем книгам автора, у которых 50 и более обзоров - 4.288462
<br>
5) Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24

На основании данного проекта можно формировать подборки для пользователей, формировать рейтинг издательств и книг, делать предположения о среднем рейтинге и количествах ревью по книгам и издательствам. Мы также отметили автора J.K. Rowling/Mary GrandPré и издательство Penguin Books, их можно однозначно добавлять в подборки для разных категорий пользователей.